libs/execution_context/ebpf_maps.c

```diff
@@ -643,28 +725,37 @@ _update_array_map_entry_with_handle(
     if (value_handle != (uintptr_t)ebpf_handle_invalid) {
         result = _validate_map_value_object(object_map, value_type, value_object);
         if (result != EBPF_SUCCESS) {
+            EBPF_LOG_MESSAGE_NTSTATUS(
+                EBPF_TRACELOG_LEVEL_ERROR, EBPF_TRACELOG_KEYWORD_ERROR, "Object validation failed.", result);
             goto Done;
         }
     }
 
-    // Release the reference on the old ID stored here, if any.
-    uint8_t* entry = &map->data[*key * map->ebpf_map_definition.value_size];
-    ebpf_id_t old_id = *(ebpf_id_t*)entry;
-    if (old_id) {
-        ebpf_assert_success(ebpf_object_dereference_by_id(old_id, value_type));
-    }
+    uint8_t* value = _get_array_map_value(map, index);
+    __analysis_assume(value != NULL);
+
+    // If this entry was in use, we need to release the reference on the old weak reference stored with this value.
+    _clear_map_entry(map, value);
 
-    ebpf_id_t id = value_object ? value_object->id : 0;
+    // Get a weak ref for the incoming object (if valid) and store it alongside the id. Note that this could be an
```

Comment:
Why?  If all IDs are weak references then you should just need to store the ID and have the weak reference be in _ebpf_id_table[index].object.   Then you don't need supplemental values as far as I know.   Back when I first implemented map-in-map and prog-array I had supplemental values for the strong references and then cleaned that up since just storing the ID was simpler, saved memory, and there may have been some other compatibility issue I ran into (I don't remember offhand).